src/lib.rs

```diff
@@ -282,12 +282,35 @@ pub extern "C" fn setsockopt(
     optval: *const c_void,
     optlen: socklen_t,
 ) -> c_int {
-}
+}*/
 
 #[no_mangle]
-pub extern "C" fn listen(fd: c_int, n: c_int) -> c_int {}
+pub extern "C" fn listen(fd: c_int, n: c_int) -> c_int {
+    let lock = match INDEX.get(fd) {
+        Some(s) => s,
+        None => return next!(listen(fd, n)),
+    };
 
-#[no_mangle]
+    let mut sock = lock.write().unwrap();
+
+    match *sock {
+        Socket::Bound(p) => (),
+        _ => return error(libc::EBADFD, -1),
+    }
+
+    if next!(listen(fd, n)) == -1 {
+        return -1;
+    }
+
+    *sock = match *sock {
+        Socket::Bound(p) => Socket::Listening(p),
+        _ => return error(libc::EBADFD, -1),
+    };
+
+    0
```

Comment:
Don't use two match statements. Put lines 301-310 inside the match statement on line 297.